AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -796,17 +798,142 @@ - (BOOL)shouldCalculateLayoutWithConstrainedSize:(ASSizeRange)constrainedSize pa
   return _calculatedDisplayNodeLayout->isValidForConstrainedSizeParentSize(constrainedSize, parentSize) == NO;
 }
 
-- (ASLayoutElementType)layoutElementType
+// TODO: coalayout: Deprecate API instead call -[ASDisplayNode setNeedsLayout]
+- (void)invalidateCalculatedLayout
 {
-  return ASLayoutElementTypeDisplayNode;
+  [self __setNeedsLayout];
 }
 
-- (BOOL)canLayoutAsynchronous
+/**
+ * Invalidate the current calculated layout so it can be recalculated in the next layout pass
+ */
+- (void)__setNeedsLayout
 {
-  return !self.isNodeLoaded;
+  ASDN::MutexLocker l(__instanceLock__);
+
+  _calculatedDisplayNodeLayout->invalidate();
+  _pendingDisplayNodeLayout->invalidate();
 }
 
-#pragma mark - Automatic Hierarchy
+/**
+ * Returns a BOOL indicating whether the layer has been marked as needing a layout update.
+ */
+- (BOOL)__needsLayout
+{
+  return _calculatedDisplayNodeLayout->isDirty();
+}
+
+/**
+ * The node's supernodes are traversed until a ancestor node is found that does not require layout. Then layout
+ * is performed on the entire node-tree beneath that ancestor
+ */
+- (void)__layoutIfNeeded
+{
+  ASDisplayNodeAssertThreadAffinity(self);
+  
+  __instanceLock__.lock();
+  
+  // Invalidate the current calculated layout so in the next layout pass it get's recalculated
+  [self invalidateCalculatedLayout];
+  
+  // Walk the node's supernodes until a ancestor node is found that does not require layout
+  if ([_supernode __needsLayout]) {
+    ASDisplayNode *supernode = _supernode;
```

Comment:
Should we create the local supernode pointer outside the if?  I'm not totally sure if this would be net-better (since it would add some retain / release calls, but would reduce weak accesses in the case the branch is taken).  Not critical either way.